### PR TITLE
Modularize passStyleOf

### PR DESCRIPTION
**Pull request by erights**
Pure refactor: A major internal refactoring, to set things up for upcoming big changes. But this refactor by itself should have no observable effects.

Well, except for the following, which affects no current code:

Before this PR, in order to allow malformed errors to be passed through marshal, we had `passStyleOf` validate them as `copyError`, which could therefore be stores within passable structures. The allowed malformedness included the possibility of extra properties leading to surprising power. These would not be passed, so there was no distributed security problem. But the possibility of these extra properties on a Passable creates a local hazard. In this PR, malformed errors are passable in the same way, but are no longer considered Passable.

Well formed errors are still Passable. However errors are no longer Comparable. They cannot be compared using sameStructure.

I have separately verified that this PR in compatible with #3525

### CODE REVIEW

File: packages/marshal/src/passStyleOf.js [hunk 1]
```diff
@@ -3,598 +3,197 @@
...
- * @param {Passable} val
- * @returns {PassStyle}
- */
-export const passStyleOf = val =>
-  // Even when a WeakSet is correct, when the set has a shorter lifetime
-  // than its keys, we prefer a Set due to expected implementation
-  // tradeoffs.
-  passStyleOfRecur(val, new Set());
-harden(passStyleOf);
+harden({ everyPassable: everyPassableChild });
```
Comment (michaelfig) [inline packages/marshal/src/passStyleOf.js]: This looks like leftovers from a code refactor.

```suggestion
harden(everyPassableChild);
```

File: packages/marshal/src/structure.js [hunk 2]
```diff
@@ -0,0 +1,215 @@
...
+    }
+  }
+};
+
+// Like the passStyleCache. An optimization only. Works because comparability
+// is guaranteed stable.
+const structureCache = new WeakMap();
+
+/**
+ * If `passable` is not a Passable, throw. Otherwise say whether it is
```
Comment (michaelfig) [inline packages/marshal/src/structure.js]: I don't think this throws anymore.  Looks like it returns `true` or `false`.

File: packages/marshal/src/structure.js [hunk 3]
```diff
@@ -0,0 +1,215 @@
...
+    }
+    default: {
+      assert.fail(X`Unexpected passStyle: ${q(passStyle)}`, TypeError);
+    }
+  }
+};
+harden(fulfillToStructure);
+
+/**
+ * This internal recursion may assume that `left` and `right` and
```
Comment (michaelfig) [inline packages/marshal/src/structure.js]:
```suggestion
 * This internal recursion may assume that `left` and `right` are
```

**Review by michaelfig — APPROVED** [review]
Looks clearly like a useful refactoring.  Please consider my comments as the only peculiarities I observed.

File: packages/marshal/src/helpers/internal-types.js [hunk 4]
```diff
@@ -0,0 +1,45 @@
+// @ts-check
+
+// eslint-disable-next-line spaced-comment
+/// <reference path="../extra-types.d.ts" />
+
+import '../types.js';
```
[the remark this answers is not in the document]
Comment (erights) [reply packages/marshal/src/helpers/internal-types.js:6]: So I should leave this code as is?

File: packages/marshal/src/helpers/copyArray.js [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+        TypeError,
+      );
+      assert(
+        desc.enumerable,
+        X`Array elements must be enumerable: ${q(i)}`,
+        TypeError,
+      );
+    }
+    assert(
+      ownKeys(descs).length === len + 1,
```
[the remark this answers is not in the document]
Comment (erights) [reply packages/marshal/src/helpers/copyArray.js]: Done. I factored out some of the common property testing logic, making this additional check pleasant.

File: packages/marshal/src/helpers/error.js [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+import '@agoric/assert/exported.js';
+import { assertChecker } from './passStyleHelpers.js';
+
+const { details: X } = assert;
+const { getPrototypeOf, getOwnPropertyDescriptors } = Object;
+const { ownKeys } = Reflect;
+
+// TODO: Maintenance hazard: Coordinate with the list of errors in the SES
+// whilelist. Currently, both omit AggregateError, which is now standard. Both
+// must eventually include it.
```
[the remark this answers is not in the document]
Comment (erights) [reply packages/marshal/src/helpers/error.js:22]: A bigger topic. Filed https://github.com/Agoric/agoric-sdk/issues/3588 to keep track of this.

File: packages/marshal/src/structure.js (hunk 2 above)
Comment (erights) [reply packages/marshal/src/structure.js]: Good catch! Yes.

**Pull request merged**